pkg/view/manager.go

```diff
@@ -0,0 +1,132 @@
+package view
+
+import (
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-faster/errors"
+	"github.com/rs/zerolog/log"
+	"html/template"
+	"io/fs"
+	"net/http"
+	"path/filepath"
+)
+
+type TemplateHandler func(w http.ResponseWriter, r *http.Request) any
+
+type TemplateRoute struct {
+	name    string
+	handler func(w http.ResponseWriter, r *http.Request) any
+	tmpl    *template.Template
+}
+
+type Manager struct {
+	templates map[string]*TemplateRoute
+	tfs       fs.FS
+	watcher   *fsnotify.Watcher
+}
+
+func NewManager(fsys fs.FS) (*Manager, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create file watcher")
+	}
+
+	m := &Manager{
+		templates: make(map[string]*TemplateRoute),
+		tfs:       fsys,
+		watcher:   watcher,
+	}
+	err = m.loadTemplates()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load templates")
+	}
+
+	// TODO breadchris setup fs different in dev for live reload
+	//go m.watchTemplateChanges()
+	return m, nil
+}
+
+func (s *Manager) loadTemplates() error {
+	return fs.WalkDir(s.tfs, "pages", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		return s.registerTemplate(path)
+	})
+}
+
+func (s *Manager) watchTemplateChanges() {
+	for {
+		select {
+		case event, ok := <-s.watcher.Events:
+			if !ok {
+				return
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				err := s.registerTemplate(event.Name)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to register template")
+				}
+			}
+		case err, ok := <-s.watcher.Errors:
+			if !ok {
+				return
+			}
+			log.Error().Err(err).Msg("watcher error")
+		}
+	}
+}
+
+func (s *Manager) registerTemplate(p string) error {
+	relPath, err := filepath.Rel("pages", p)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get relative path for %s", p)
+	}
+
+	// split file name from path
+	b := filepath.Base(relPath)
+	if b == "index.html" {
+		relPath = filepath.Dir(relPath)
+	} else {
+		relPath = relPath[:len(relPath)-len(filepath.Ext(relPath))]
+	}
+
+	urlPath := "/" + filepath.ToSlash(relPath)
+	if relPath == "." {
+		urlPath = "/"
+	}
+
+	tmpl, err := template.ParseFS(s.tfs, "layout/base.html", p)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse template %s", p)
+	}
+
+	s.templates[urlPath] = &TemplateRoute{
+		name:    relPath,
+		handler: nil,
+		tmpl:    tmpl,
+	}
+
+	// TODO breadchris need setup fs different in dev for live reload
+	//return s.watcher.Add(path)
+	return nil
+}
+
+func (s *Manager) BuildRouter() (*chi.Mux, error) {
```

Comment:
I'm not totally sure how any of this works as there is a lot of autowiring happening.

I want there to be a function that looks like this: https://github.com/go-gitea/gitea/blob/main/routers/web/web.go#L405

Where do we define the logic for a route?

Reply:
I knew you were going to pick up on this one ;) I was coding this one a little more for fun than what I think should be in here right now. I agree that we should have it look like gitea's for simplicity.

We do need logic setup so that we can load templates from disk during development and then from the embed in prod (ex. config option). This will help speed up local dev significantly as UI changes are immediate.

Reply:
I was playing with the feeling of being able to just drop HTML files in this directory and having something happen. 

![Screenshot 2024-03-26 at 9 48 52 AM](https://github.com/scratchdata/scratchdata/assets/1069294/43a67651-2fd4-49a2-9b28-a0d215d14fee)